- Added a new trainer randomisation option "Type Match" that replaces…

Humg12 opened the pull request:
… trainer pokemon with one that shares at least one type with it.

Commit:
- Added a new trainer randomisation option "Type Match" that replaces trainer pokemon with one that shares at least one type with it.

Humg12 commented:
I've never coded in java before, so probably worth checking carefully, but I generated ~10 roms and it seemed to be working as intended.

KittyPBoxx commented:
Just so I understand, this is a mode, similar to 'TypeThemed' except, instead of areas having a random type, each trainer has it's type based of the type of it's first pokemon in an un-randomized game? 

If so, I think I'll need to add a check in pickReplacement to ensure that if the player is using a custom ban list, it does have at least 1 pokemon of that type.

Humg12 commented:
> Just so I understand, this is a mode, similar to 'TypeThemed' except, instead of areas having a random type, each trainer has it's type based of the type of it's first pokemon in an un-randomized game?
> 
> If so, I think I'll need to add a check in pickReplacement to ensure that if the player is using a custom ban list, it does have at least 1 pokemon of that type.

Close, but not just the first pokemon. Every pokemon is replaced base on their type. So if a trainer had a wingull and a geodude. The wingull would be replaced with either a random water type or a random flying type, and the geodude with a random rock or a random ground type.

Though you're right that it should have a check about the custom ban list to make sure that there's at least one pokemon of the type.

Commit:
- Added a check to ensure there's a valid pokemon of the chosen type
- Added a note to the tooltip explaining it will be completely random if there are no valid types.

Humg12 commented:
I've added that safety check in pick replacement. Now if it can't find a valid pokemon for the type it will just add all pokemon back to the potential pool. Also adjusted the tooltip for the mode to explain that.

Humg12 commented:
Also, I finished a run with my friend with this setting on, and it worked smoothly for the most part, but there was a single trainer that caused my game to freeze when initiating a fight with her (the spinner next to the berry master's house). Not 100% sure it's related to my change, and unfortunately I didn't generate a log, so I don't know what pokemon she was meant to have that could have been causing it. Luckily she was an optional, so I just skipped her.
